### DISCUSSION TITLE
Cannot upgrade from 9.12.1 to 9.12.2 (or any other later version)

### DISCUSSION THREAD

**Discussion by wilkomanger** [directus/directus#14347] [Archive]
### Describe the Bug

Hi, thanks for all the work done with Directus!

Unfortunately when upgrading from 9.12.1 -> 9.12.2, or any other later version than that, Directus does not seem to boot. When I run `docker-compose logs directus`, all I get is:
```
Attaching to directus
directus    | 
```
So it's not clear to me what's wrong. Are there any breaking changes between these releases?

Directus is deployed through Docker on Ubuntu 21.10.

### To Reproduce

1. Setup Directus 9.12.1 using docker-compose
2. Upgrade to 9.12.2
3. See empty logs and Directus not starting

### What version of Directus are you using?

9.12.1

### What version of Node.js are you using?

Packaged with Docker

### What database are you using?

Postgres 13

### What browser are you using?

Firefox

### How are you deploying Directus?

Docker

**Comment by rijkvanzanten** [comment]
There's no known breaking changes that are prevention server startup. It sounds like your docker is having some trouble attaching to the process to stream the logs? Maybe try checking through the docker gui to see if that's able to show some logs?

**Reply by wilkomanger** [reply]
Haven't used the the Docker GUI yet, but when I run `docker-compose logs -f`, I see this additional line:
```
directus exited with code 243
```
Does this help? I saw that in relation to node/npm it's probably an `EACCESS` error?

**Reply by rijkvanzanten** [reply]
Looks it might be a known issue that others are running into in the combination Node + Docker? Not sure! https://stackoverflow.com/questions/72855226/i-have-error-with-docker-compose-exited-with-code-243

**Reply by wilkomanger** [reply]
Maybe, thanks! Looking at the Directus Dockerfile, am I correct that adding `NODE_VERSION: '16.14.2'` under `environment` in docker-compose.yaml should change the Node version? Or do I have to use my own Dockerfile? Not a Docker expert :sweat_smile:

**Reply by rijkvanzanten** [reply]
I ehh.. have no idea :C I'm not too well versed in Docker magic either

**Comment by azrikahar** [comment]
I'd also agree that there isn't any notable changes between those 2 versions that will cause such issue (and there aren't similar reports either), but perhaps you can try with a clean Directus instance, with no bind mounts or volumes at all, and see whether it works? Just so we can rule out any other potential culprits.

A minimal compose you can try may be:

```yaml
services:
    directus:
        image: directus/directus:9.14.1 # or just "latest" if that's preferred
        container_name: directus
        restart: unless-stopped
        ports:
            - 8055:8055 # feel free to it to another port if necessary
        environment:
            - KEY=directus-key
            - SECRET=directus-secret
            - ADMIN_EMAIL=admin@example.com
            - ADMIN_PASSWORD=password
```

**Reply by Logrythmik** [reply]
I am getting the same exact issue in Kubernetes. It's definitely not something to do with Docker Desktop. I too get zero log entries.

Our DB is MySQL: 5.7

I've tried upgrading to 9.12.2 and up with no success or output of any kind.

**Reply by azrikahar** [reply]
Have you tried it with the minimal setup so that we can attempt to narrow it down? I'm not aware of any big changes in `9.12.2` specifically beyond one fix that's meant to fix Flows triggering multiple times, so I'm not sure what could be the cause here.

**Reply by LevinGermann** [reply]
@azrikahar Not sure if this helps, but we got the same problem while trying to upgrade from 9.11.1 to the latest 9.14.1.
We run Directus in Kubernetes, and the container just crashes with no output with the exit code 243.
(LOG_LEVEL = debug. No mounts or external volumes. Storage location set to GCS.)

Taking @Logrythmik 's comment into consideration, an upgrade to 9.12.1 works. The next "small" update to 9.12.2 gives the error code 243.
I'm able to start the server with our database if I run it directly from source.

**Reply by LevinGermann** [reply]
https://github.com/nodejs/docker-node/issues/1734
does this sound related?

**Reply by wilkomanger** [reply]
@azrikahar I've tried your minimal example, that seems to work :thinking: Still, this bug only happens specifically with >=9.12.2. Is it possible that the >=9.12.2 Dockerfile is built with a specific Node version that has this issue?

**Reply by azrikahar** [reply]
> [nodejs/docker-node#1734](https://github.com/nodejs/docker-node/issues/1734) does this sound related?

@LevinGermann Nice catch! It may very well be related, though I do wonder what are the specific configs that are causing this, since @wilkomanger confirmed the minimal example works. You did mention no mounts or external volumes, but are there any other environment variables you may think that could affect this? or perhaps the minimal example doesn't work at all in Kubernetes? 🤔

> @azrikahar I've tried your minimal example, that seems to work 🤔 Still, this bug only happens specifically with >=9.12.2. 

@wilkomanger Thanks for confirming! As the minimal example worked, could this be config/env-variables related somehow? 

> Is it possible that the >=9.12.2 Dockerfile is built with a specific Node version that has this issue?

It is very much possible. @LevinGermann linked a very plausible issue with `v16.15.1`, and since the current Dockerfile targets `16-alpine`:

https://github.com/directus/directus/blob/b1585e0556dfc6596105a9ef5336c6880dc15e65/docker/Dockerfile#L1

I assume it's possible that it just so happens to coincide with roughly the same release dates between Node and Directus.

I'm still wondering if it's a very specific element causing this though, as the minimal example does work as you mentioned. So it feels like it'd be great if we can pinpoint what it is precisely.

---

Seems like `9.12.2` Docker image is indeed on Node 16.15.1 as Directus 9.12.1 was released on 10 June, whereas [Node 16.15.1 alpine image (i.e. 16-alpine)](https://hub.docker.com/layers/node/library/node/16.15.1-alpine/images/sha256-9da65f99264be2a78682095c4789b3d8cab12e0012def7d937d7125ed6e7695c?context=explore) was released on 8 June.

This is not ideal @LevinGermann @wilkomanger, but would you guys mind trying this sanity check of making it 16.15.0 for us to be absolute sure 16.15.1 is the culprit? I assume the following quick Dockerfile should work as our experiment (feel free to change/remove some of the lines, mainly just need to make sure it's Node 16.15.0):

```Dockerfile
### Base
# Feel free to use 9.14.1 as well
FROM directus/directus:9.12.2 as base

### Node 16.15.0
FROM node:16.15.0-alpine

ENV \
  DB_CLIENT="sqlite3" \
  DB_FILENAME="/directus/database/database.sqlite" \
  EXTENSIONS_PATH="/directus/extensions" \
  STORAGE_LOCAL_ROOT="/directus/uploads"

RUN \
  # Upgrade system and install 'ssmtp' to be able to send mails
  apk upgrade --no-cache && apk add --no-cache \
  ssmtp \
  # Add support for specifying the timezone of the container
  # using the "TZ" environment variable.
  tzdata

USER node
WORKDIR /directus

COPY --from=base /directus .

CMD npx directus bootstrap && npx directus start
```

**Reply by LevinGermann** [reply]
Thanks @azrikahar . I built a custom (9.14.1) Directus release using 16.15.0-alpine as the node version: 
https://github.com/Explorial/directus/blob/fixDockerError/docker/Dockerfile#L1

If I check the node version, it seems correct:
`docker run -it --rm f9b5fe72dfee /bin/sh -c 'node --version'`
-> v16.15.0

And: Tadah!
![image](https://user-images.githubusercontent.com/9347206/178928691-28c6eb81-8a75-4d16-9c99-39a28a30fba3.png)
![image](https://user-images.githubusercontent.com/9347206/178929127-5be4d7f2-4ce6-40a7-8239-ffbd093b7479.png)

So it really seems like that the node version is the culprit, although I still have no idea why.

**Reply by azrikahar** [reply]
@LevinGermann Thanks for helping to confirm this on your end. This really seems rather bizarre... I'll try to check out the linked issue in the docker-node repo, but it does seems to be pretty tricky to pinpoint which piece of code in Directus that is unfortunately triggering this inconspicuous bug in docker node v16.15.1. 

Will try to bring this up to the team to see what would be the best course of action here (obvious answer most likely would be pinning it to `16.15.0`, but perhaps more tests would be for the best to see if this is also present in newer versions such as docker node 18 for example)

**Reply by Logrythmik** [reply]
Thank you everyone!

**Reply by csageder** [reply]
I can confirm too, that the above solution works in our kubernetes environment too. Just tested it with directus 9.14.1 and node 16.15.0 - there is an issue in node for kubernetes [https://github.com/nodejs/docker-node/issues/1737](https://github.com/nodejs/docker-node/issues/1737)

**Comment by wilkomanger** [comment]
I noticed while running 9.12.1 that I had some read/write errors for the extension folder. Fixing these permissions, I was able to upgrade to 9.14.1 :tada:

Of course, it would be nice that these permission errors don't crash Directus as a whole :smile: Hopefully Node fixes this soon

**Reply by wilkomanger** [reply]
I see you have the same error in your logs @LevinGermann, would be worth while to fix it and try upgrading

**Reply by azrikahar** [reply]
> I noticed while running 9.12.1 that I had some read/write errors for the extension folder. Fixing these permissions, I was able to upgrade to 9.14.1 🎉

Glad it's resolved! This does remind me of https://github.com/directus/directus/discussions/9955#discussioncomment-1677445 and https://github.com/directus/directus/discussions/12160#discussioncomment-2366804, though it shouldn't ever crash the container in such scenario 🤔 might be a combination of both that and the docker node 16.15.1 bug of course.

**Reply by LevinGermann** [reply]
Thanks for the input @wilkomanger !
Since we don't use extensions (yet), I added now a simple emtyDir to our k8s deployment:

```
...
spec:
  template:
    spec:
      containers:
        ...
        - name: directus
          ...
          volumeMounts:
            - mountPath: /directus/extensions
              name: extensions-volume
      volumes:
        - name: extensions-volume
          emptyDir: {}
```

This solved the permission error on startup:
![image](https://user-images.githubusercontent.com/9347206/179579012-3b4650bb-61d3-41c9-9812-57344944a42b.png)

But unfortunately the server still crashes with error 243 if I use the official docker image of v9.14.1 :-(
Nevertheless, thanks again for the info!
@azrikahar
